코드 수정

sangyun5108 opened the pull request:
기존에 조건부 렌더링을 통해 income,expediture,incomeExpediture 리스트를 시각화 했던 부분을
수정했습니다. 윤섭님께서 말씀해주신 부분에 대해 제가 정확히 수정을 했는지 피드백 부탁드립니다.
질문사항: map메소드를 사용해서 여러 li태그를 return 할때 key라는 속성을 필요로한다고 하는데, 저는 계속 index를 key값으로 부여를 해줬습니다. 혹시 이 방법이 맞는 방법인가요..? 아니면 윤섭님,종호님께서 사용하셨던 방법을 알려주시면 감사하겠습니다.

sangyun5108 committed:
코드 수정

기존에 조건 렌더링 했던 부분을
수정했습니다. 윤섭님께서 말씀해주신 부분에 대해 제가 정확히 수정을 했는지 피드백 부탁드립니다

SimYunSup commented on src/components/expediture.js:
```diff
@@ -10,6 +10,7 @@ const Expediture = ({expediture}) => {
     const nowmonth = date.getMonth()<10?`0${date.getMonth()+1}`:`${date.getMonth()}`;
     const nowdate = date.getDate()<10?`0${date.getDate()}`:`${date.getDate()}`;
     const months = ['Jan','Feb','Mar','Apr','May','Jun','July','Aug','Sep','Oct','Nov','Dec'];
+    let id = 100;
```
함수 컴포넌트와 hooks에 대한 공부가 더 필요할 것 같습니다! 만약 Expediture 컴포넌트가 재렌더링된다면 id가 초기화되기 때문에 id를 만들어 놓고 싶다면 바깥에 두거나 localstorage에 두어야겠죠?

SimYunSup commented on src/components/income.js:
```diff
@@ -10,6 +10,7 @@ const Income = ({income}) => {
     const nowmonth = date.getMonth()<10?`0${date.getMonth()+1}`:`${date.getMonth()}`;
     const nowdate = date.getDate()<10?`0${date.getDate()}`:`${date.getDate()}`;
     const months = ['Jan','Feb','Mar','Apr','May','Jun','July','Aug','Sep','Oct','Nov','Dec'];
+    let id = 0;
```
Expediture 컴포넌트와 같습니다.

SimYunSup commented on src/components/showList.js:
```diff
@@ -29,27 +24,35 @@ const ShowList = ({lists}) => {
+                <h3>{type}</h3>
+                {newLists.map((list,index)=>{
+                    return(
+                        <li key={index}>{list.label} {list.amount}</li>
```
보통 key를 index로 두지 않습니다. React가 리스트 렌더링을 할 때 key를 기반으로 렌더링하기 때문에 index를 key로 두지 않고 고유(unique) ID를 key로 사용합니다.

SimYunSup replied:
https://robinpokorny.medium.com/index-as-a-key-is-an-anti-pattern-e0349aece318

SimYunSup commented on src/components/showList.js:
```diff
@@ -29,27 +24,35 @@ const ShowList = ({lists}) => {
         }
     }
 
+    const makeList = (type) => {
```
굳이 함수로 만들 필요가 있나요? return 함수에 합쳐놓거나 컴포넌트를 분리하는 것이 깔끔한 코드가 될 것 같습니다.

SimYunSup commented on src/components/expediture.js:
```diff
@@ -62,7 +64,7 @@ const Expediture = ({expediture}) => {
 const mapDispatchToProps = (dispatch) => {
     return {
-        expediture: (amount,label,kind,year,month,date) => dispatch(expediture(amount,label,kind,year,month,date))
+        expediture: (amount,label,kind,year,month,date,id) => dispatch(expediture(amount,label,kind,year,month,date,id))
```
useDispatch를 써주세요 :)

SimYunSup reviewed (commented):
함수 컴포넌트에서 return 문 밖에서 렌더링 하는 함수를 만드는 것은 지양해야 합니다. 처음에 구조를 보고 파악하기 힘들어지니까요.

sangyun5108 committed:
코드 수정

피드백 받은 코드
리랜더링시 초기화되는 id값 함수밖선언
useDispatch,useSelector사용
makeList를 컴포넌트로 분리
key값을 index가 아닌 고유 id로 부여

SimYunSup commented on src/components/income.js:
```diff
@@ -42,7 +46,8 @@ const Income = ({income}) => {
         e.preventDefault();
         const exchangeMonth = months[Number(inputMonth)-1];
-        income(amount,label,KIND,inputYear,exchangeMonth,Number(inputDate));
+        incomeDispatch(income(amount,label,KIND,inputYear,exchangeMonth,Number(inputDate),id));
+        id++;
```
income이라는 함수에서 redux action에 대한 type을 정의하고 있는데 KIND가 따로 들어가는 이유라도 있나요?

SimYunSup commented on src/components/showList.js:
```diff
@@ -31,25 +29,25 @@ const ShowList = ({lists}) => {
                 </>
-            )
+            );
         }
```
blueBtn과 redBtn 상태로 두는게 아닌, type상태를 두고 해당 상태를 income/expediture/null로 두고 MakeList에 넘기는게 더 효율적입니다. 그리고 댓글에서도 말했지만 함수 컴포넌트에서 return 문 밖에서 렌더링 하는 함수를 만드는 것은 지양해야 합니다.

SimYunSup commented on src/redux/reducers.js:
```diff
@@ -8,7 +8,8 @@ const initialState = {
             kind:'',
             year:'',
             month:'',
-            date:''
+            date:'',
+            id:0
```
InitialState에는 내용이 없거나 의미있는 내용이 들어가야겠죠? :)

SimYunSup reviewed (changes requested):
이외에도 year, month, date에 대한 입력을 제한해주셔야 합니다. 다른 지적할 부분은 다음 코드리뷰때 해드릴게요 :)

그리고 커밋을 한 번에 하는 것이 아닌, 의미있는 변경 단위로 해주셔야 합니다.

sangyun5108 committed:
Relocation ID

Dispatch를 하면서 부여해주었던 Id값을
action을 생성해주는 함수에서 action이 생성이 될때마다, 선부여 후 증가 방식으로 구현

sangyun5108 committed:
remove kind

kind라는 key값을 통해서 income인지 expediture인지 구분하는 것이 아닌,
action의 type을 통해서 구분해주는 것으로 수정했습니다.

sangyun5108 committed:
edit initialState

initialState가 갖고 있던 의미없는 key값들삭제

sangyun5108 committed:
edit function

함수 컴포넌트의 return 문 밖에서 랜더링해주는 부분을 제거하고, return문에서 랜더링 하도록 구현, 랜더링시 비용이 많이 드는 sumIncome, sumExpediture function을
ShowList function 밖에 선언해줌으로써, 필요할 때만 사용할 수 있도록 구현
